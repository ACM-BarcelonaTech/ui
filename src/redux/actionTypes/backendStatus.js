const BACKEND_STATUS = 'backendStatus';

/**
 * Starts loading pipeline status from the API.
 */
const BACKEND_STATUS_LOADING = `${BACKEND_STATUS}/backendStatusLoading`;

/**
 * Loaded pipeline status from the API.
 */
const BACKEND_STATUS_LOADED = `${BACKEND_STATUS}/backendStatusLoaded`;

/**
 * Backend status update was received.
 */
const BACKEND_STATUS_UPDATED = `${BACKEND_STATUS}/backendStatusUpdated`;

/**
 * Error state when backend status could not be loaded from the API.
 */
const BACKEND_STATUS_ERROR = `${BACKEND_STATUS}/backendStatusError`;

export {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_UPDATED,
  BACKEND_STATUS_ERROR,
};
