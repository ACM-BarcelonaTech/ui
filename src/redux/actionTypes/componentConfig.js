/**
 * Sets plot configs and data.
 */
const COMPONENT_CONFIG = 'componentConfig';
const LOAD_CONFIG = `${COMPONENT_CONFIG}/load`;
const UPDATE_CONFIG = `${COMPONENT_CONFIG}/update`;
const SAVE_CONFIG = `${COMPONENT_CONFIG}/save`;
const PLOT_DATA_LOADED = `${COMPONENT_CONFIG}/plotDataLoaded`;
const PLOT_DATA_LOADING = `${COMPONENT_CONFIG}/plotDataLoading`;

export {
  LOAD_CONFIG, UPDATE_CONFIG, SAVE_CONFIG,
  PLOT_DATA_LOADED, PLOT_DATA_LOADING,
};
