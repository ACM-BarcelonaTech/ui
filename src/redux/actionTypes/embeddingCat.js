/**
 * Sets embedding categorical config.
 */
const EMBEDDING_CAT_CONFIG = 'embeddingCatConfig';
const EMBEDDING_CAT_CONFIG_UPDATE = `${EMBEDDING_CAT_CONFIG}/update`;


export default EMBEDDING_CAT_CONFIG_UPDATE;
