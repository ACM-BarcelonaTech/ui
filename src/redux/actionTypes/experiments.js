const EXPERIMENTS = 'experiments';

/**
 * Create experiments
 */
const EXPERIMENTS_CREATED = `${EXPERIMENTS}/created`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADING = `${EXPERIMENTS}/loading`;

/**
 * Create experiments
 */
const EXPERIMENTS_LOADED = `${EXPERIMENTS}/loaded`;

/**
 * Experiments updated
 */
const EXPERIMENTS_UPDATED = `${EXPERIMENTS}/updated`;

/**
 * Delete experiments
 */
const EXPERIMENTS_DELETED = `${EXPERIMENTS}/deleted`;

/**
 * Saving experiment
 */
const EXPERIMENTS_SAVING = `${EXPERIMENTS}/saving`;

/**
 * Experiment saved
 */
const EXPERIMENTS_SAVED = `${EXPERIMENTS}/saved`;

/**
 * Update experiments
 */
const EXPERIMENTS_ERROR = `${EXPERIMENTS}/error`;

/**
 * Data download is requested
 */
const EXPERIMENTS_DATA_DOWNLOAD_REQUESTED = `${EXPERIMENTS}/downloadRequested`;

/**
 * Download link is ready
 */
const EXPERIMENTS_DATA_DOWNLOAD_READY = `${EXPERIMENTS}/downloadReady`;

/**
 * Download link is clicked
 */
const EXPERIMENTS_DATA_DOWNLOADED = `${EXPERIMENTS}/downloaded`;

/**
 * There is download error
 */
const EXPERIMENTS_DATA_DOWNLOAD_ERROR = `${EXPERIMENTS}/downloadError`;

export {
  EXPERIMENTS_CREATED,
  EXPERIMENTS_UPDATED,
  EXPERIMENTS_ERROR,
  EXPERIMENTS_LOADING,
  EXPERIMENTS_LOADED,
  EXPERIMENTS_DELETED,
  EXPERIMENTS_SAVING,
  EXPERIMENTS_SAVED,
  EXPERIMENTS_DATA_DOWNLOAD_REQUESTED,
  EXPERIMENTS_DATA_DOWNLOAD_READY,
  EXPERIMENTS_DATA_DOWNLOADED,
  EXPERIMENTS_DATA_DOWNLOAD_ERROR,
};
