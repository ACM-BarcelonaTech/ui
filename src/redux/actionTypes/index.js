const CELLS = 'CELLS';
const LOAD_CELLS = `${CELLS}.LOAD`;

const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const DIFF_EXPR = 'DIFF_EXPR';
const LOAD_DIFF_EXPR = `${DIFF_EXPR}.LOAD`;
const UPDATE_DIFF_EXPR = `${DIFF_EXPR}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const UPDATE_CELL_INFO = 'UPDATE_CELL_INFO';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_CELLS,
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
  UPDATE_CELL_INFO, SET_FOCUSED_GENE,
};
