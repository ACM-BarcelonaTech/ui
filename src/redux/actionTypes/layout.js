/* eslint-disable import/prefer-default-export */
const UPDATE_LAYOUT = 'UPDATE_LAYOUT';

export { UPDATE_LAYOUT };
