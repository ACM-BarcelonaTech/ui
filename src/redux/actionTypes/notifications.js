const NOTIFICATIONS = 'notifications';

/**
 * Creates a message to be pushed to the user.
 */
const NOTIFICATIONS_PUSH_MESSAGE = `${NOTIFICATIONS}/pushMessage`;

/**
 * Clears the specified notification.
 */
const NOTIFICATIONS_CLEAR_MESSAGE = `${NOTIFICATIONS}/clearMessage`;

export { NOTIFICATIONS_PUSH_MESSAGE, NOTIFICATIONS_CLEAR_MESSAGE };
