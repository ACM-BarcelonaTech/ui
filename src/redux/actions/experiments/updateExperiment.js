import _ from 'lodash';

import fetchAPI from 'utils/fetchAPI';
import {
  EXPERIMENTS_UPDATED, EXPERIMENTS_SAVING, EXPERIMENTS_SAVED, EXPERIMENTS_ERROR,
} from 'redux/actionTypes/experiments';
import endUserMessages from 'utils/endUserMessages';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import { isServerError, throwIfRequestFailed } from '../../../utils/fetchErrors';

const convertedToApiModel = (experiment) => {
  const {
    id = null, name = null, projectUuid = null, ...restOfExperiment
  } = experiment;

  const convertedExperiment = {
    ...restOfExperiment,
    experimentId: id,
    experimentName: name,
    projectId: projectUuid,
  };

  return _.omitBy(convertedExperiment, _.isNull);
};

const updateExperiment = (
  experimentId,
  experimentDiff,
) => async (dispatch) => {
  try {
    dispatch({
      type: EXPERIMENTS_SAVING,
    });

    const url = `/v1/experiments/${experimentId}`;
    try {
      const response = await fetchAPI(
        url,
        {
          method: 'PUT',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(convertedToApiModel(experimentDiff)),
        },
      );

      const json = await response.json();
      throwIfRequestFailed(response, json, endUserMessages.ERROR_SAVING);

      dispatch({
        type: EXPERIMENTS_UPDATED,
        payload: {
          experimentId,
          experimentDiff,
        },
      });

      dispatch({
        type: EXPERIMENTS_SAVED,
      });
    } catch (e) {
      let { message } = e;
      if (!isServerError(e)) {
        console.error(`fetch ${url} error ${message}`);
        message = endUserMessages.CONNECTION_ERROR;
      }
      dispatch({
        type: EXPERIMENTS_ERROR,
        payload: {
          error: message,
        },
      });

      pushNotificationMessage(
        'error',
        message,
      );
    }
  } catch (e) {
    pushNotificationMessage('error', endUserMessages.ERROR_SAVING);
  }
};

export default updateExperiment;
