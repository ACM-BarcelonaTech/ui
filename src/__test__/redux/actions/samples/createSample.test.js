import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { v4 as uuidv4 } from 'uuid';

import createSample from 'redux/actions/samples/createSample';
import initialSampleState from 'redux/reducers/samples/initialState';
import initialProjectState, { projectTemplate } from 'redux/reducers/projects/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

import { SAMPLES_CREATE, SAMPLES_SAVING, SAMPLES_ERROR } from 'redux/actionTypes/samples';
import updateExperiment from 'redux/actions/experiments/updateExperiment';

import pushNotificationMessage from 'utils/pushNotificationMessage';

jest.mock('../../../../redux/actions/experiments/updateExperiment');
updateExperiment.mockImplementation(() => async () => { });

jest.mock('../../../../utils/pushNotificationMessage');
pushNotificationMessage.mockImplementation(() => async () => { });

jest.mock('localforage');

enableFetchMocks();

const mockStore = configureStore([thunk]);

jest.mock('uuid');
const sampleUuid = 'abc123';
uuidv4.mockImplementation(() => sampleUuid);

const sampleName = 'test sample';

describe('createSample action', () => {
  const projectUuid = 'qwe234';
  const experimentId = 'exp234';

  const mockType = '10x Chromium';

  const mockProject = {
    ...projectTemplate,
    name: 'test project',
    uuid: projectUuid,
    experiments: [experimentId],
  };

  const mockExperiment = {
    ...experimentTemplate,
    name: 'Experiment 1',
    id: experimentId,
  };

  const initialState = {
    samples: { ...initialSampleState },
    experiments: {
      ...initialExperimentState,
      [experimentId]: mockExperiment,
    },
    projects: {
      ...initialProjectState,
      ids: [projectUuid],
      [projectUuid]: mockProject,
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());

    fetchMock.resetMocks();
    fetchMock.doMock();
  });

  it('Runs correctly', async () => {
    const store = mockStore(initialState);

    fetchMock.mockResponse(JSON.stringify({}), { url: 'mockedUrl', status: 200 });

    await store.dispatch(createSample(projectUuid, sampleName, mockType));

    // Fetch call is made
    const fetchMockFirstCall = fetchMock.mock.calls[0];

    const { body: fetchBody, method: fetchMethod } = fetchMockFirstCall[1];
    expect(fetchMockFirstCall[0]).toEqual(`http://localhost:3000/v1/projects/${projectUuid}/${mockProject.experiments[0]}/samples`);

    expect(fetchMethod).toEqual('POST');
    expect(JSON.parse(fetchBody)).toMatchSnapshot();

    // Sends correct actions
    const actions = store.getActions();

    expect(actions[0].type).toEqual(SAMPLES_SAVING);
    expect(actions[1].type).toEqual(SAMPLES_CREATE);

    // Calls update experiment on success of fetch
    expect(updateExperiment).toHaveBeenCalledWith(experimentId, { sampleIds: [sampleUuid] });
  });

  it('Shows error message when there was a fetch error', async () => {
    const store = mockStore(initialState);

    const fetchErrorMessage = 'someFetchError';

    fetchMock.mockResponse(JSON.stringify({ message: fetchErrorMessage }), { url: 'mockedUrl', status: 400 });

    // Fails with error message we sent in response to fetch
    await expect(
      store.dispatch(
        createSample(projectUuid, sampleName, mockType),
      ),
    ).rejects.toEqual(fetchErrorMessage);

    // Sends correct actions
    const actions = store.getActions();

    expect(actions[0].type).toEqual(SAMPLES_SAVING);
    expect(actions[1].type).toEqual(SAMPLES_ERROR);

    expect(pushNotificationMessage).toHaveBeenCalledWith('error', fetchErrorMessage);
  });
});
