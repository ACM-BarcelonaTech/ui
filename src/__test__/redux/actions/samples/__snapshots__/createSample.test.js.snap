// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSample action Runs correctly 1`] = `
Object {
  "complete": false,
  "createdDate": "2020-01-01T00:00:00.000Z",
  "error": false,
  "fileNames": Array [],
  "files": Object {},
  "lastModified": "2020-01-01T00:00:00.000Z",
  "metadata": Object {},
  "name": "test sample",
  "projectUuid": "qwe234",
  "type": "10x Chromium",
  "uuid": "abc123",
}
`;
